:book: Adding OLM V1 architecture document (#1301)
operator-framework/operator-controller: docs/drafts/architecture.md

## OLM V1 Architecture

> This document talks about the V1 architecture. OLM V1 consists of many components, however majority of the code is divided in to two components

Reviewer: ```suggestion
This document describes the OLM v1 architecture. OLM v1 consists of two main components:
```

* [catalogD](https://github.com/operator-framework/catalogd)

The below diagram and text explains the sub-components within each code repository and how they related to each-other.

### Operator-controller:

operator-controller is the central component of OLM v1, that consumes all of the components below to extend Kubernetes to allows users to install, and manage the lifecycle of other extensions

* Operator-controller queries the catalogd (catalogd HTTP Server) to get catalog information.  Once received the catalog information is saved  to catalog-cache. The cache will be updated automatically if a Catalog is noticed to have a different resolved image reference.
* Resolver in operator-controller helps the extension controller to filter the bundle reference after applying the user restrictions (e.g. name, priority etc) and returns the bundle reference to the extension controller.
* Cluster extension controller reaches out to the registry to download the bundle container images, saves it to the bundle cache,  unpacks it and applies the bundle manifests to the cluster.  It is also Responsible for figuring out which bundle to upgrade.

> * Bundle cache returns the cache for the bundle. If a cache does not already exist, a new one will be created.

Reviewer 2: Feels odd that we are talking about the components by name before a reader sees the diagram

Author: Fixed it in https://github.com/operator-framework/operator-controller/pull/1301/commits/67c1540f1ed18d59b3dbc31f0fb59ff2a8a95eb4

### Catalogd:

> Catalogd unpacks [file-based catalog (FBC)](https://olm.operatorframework.io/docs/reference/file-based-catalogs/#docs) content that is packaged and shipped in container images, for consumption by clients on-clusters (unpacking from other sources, like git repos, OCI artifacts etc, are in the roadmap for catalogD). As component of the Operator Lifecycle Manager (OLM) v1 microservices architecture, catalogD hosts metadata for Kubernetes extensions packaged by the authors of the extensions, as a result helping customers discover installable content.

Reviewer: I'd move `(unpacking from other sources, like git repos, OCI artifacts etc, are in the roadmap for catalogD)` down to the end of the section. It's a great call out, I just think it's misplaced here.

Author: This is resolved

> Catalogd unpacks [file-based catalog (FBC)](https://olm.operatorframework.io/docs/reference/file-based-catalogs/#docs) content that is packaged and shipped in container images, for consumption by clients on-clusters (unpacking from other sources, like git repos, OCI artifacts etc, are in the roadmap for catalogD). As component of the Operator Lifecycle Manager (OLM) v1 microservices architecture, catalogD hosts metadata for Kubernetes extensions packaged by the authors of the extensions, as a result helping customers discover installable content.

Reviewer: ```suggestion
Catalogd unpacks [file-based catalog (FBC)](https://olm.operatorframework.io/docs/reference/file-based-catalogs/#docs) content that is packaged and shipped in container images, for consumption by clients on-clusters (unpacking from other sources, like git repos, OCI artifacts etc, are in the roadmap for catalogD). It serves the extension metadata, provided by the extension authors, found in the FBC, making it possible for on-cluster clients to discover installable content.
```

* Catalogd can be broken down in to three sub-components i.e. ClusterCatalog controller, catalogd http server, catalogd content cache.
* Catalog controller  is responsible for pulling FBC based catalog images from registry and unpacking them into the catalog content cache. It is also responsible for reconciling the latest changes in the cluster catalog.
* Catalogd http server is responsible for serving catalog information to clients e.g. cluster extension controller.
* Catalogd content cache is maintained by the catalog controller and used by the catalogd http server to answer queries from clients.

> ### Diagram

Reviewer: I'd suggest moving the diagram to the top.
